### PR TITLE
Fix bytearray iteration in 0.29.x

### PR DESCRIPTION
**Pull request by Kenrick0**
Fixing #3473 
Causes the NextNode to have PyObjectType which is Coerced to unsigned char.

### CODE REVIEW

**Comment by da-woods** [comment]
It doesn't seem to break any of the existing optimizations that I can see, but it does lead to unused utility-code being imported.

If I run

```
def f(x: unicode):
    for y in x:
        pass
```

without your patch it compiles cleanly, but with your patch it gives

```
warning: ‘__Pyx__PyObject_AsPy_UCS4’ defined but not used [-Wunused-function]
 1938 | static Py_UCS4 __Pyx__PyObject_AsPy_UCS4(PyObject* x) {
```

That suggests to me that this isn't quite the right way to solve it. I don't have a more helpful suggestion at the moment though.

File: Cython/Compiler/ExprNodes.py [hunk 1]
```diff
@@ -2876,6 +2876,8 @@ def infer_type(self, env, iterator_type=None):
             iterator_type = self.iterator.infer_type(env)
         if iterator_type.is_ptr or iterator_type.is_array:
             return iterator_type.base_type
+        elif iterator_type is py_object_type:
+            return py_object_type
```
Comment (scoder) [inline Cython/Compiler/ExprNodes.py:2880]: I think we must look at `iterator_type.sequence` here.
I'm ok with hacking this in for 0.29.x right in `NextNode`, but only specifically for `bytearray_type`. The master branch has a proper fix.

Comment (Kenrick0) [reply Cython/Compiler/ExprNodes.py:2880]: The relevant part of the parse tree after `AnalyseExpressionsTransform` without the patch:
```python
- stats[1]: ForInStatNode(pos=(test.py:3:4))
  - target: NameNode(type=<CNumericType unsigned char>, name="elem")
  - item: NextNode(type=<CNumericType unsigned char>)
  - iterator: IteratorNode(type=<PyObjectType>)
    - sequence: NameNode(type=<PyByteArray_Type>, name="data")
  - body: StatListNode(pos=(test.py:4:8))
    - stats[0]: PrintStatNode(pos=(test.py:4:8))
      - arg_tuple: TupleNode(type=<PyTuple_Type>)
        - args[0]: CoerceToPyTypeNode(type=<PyObjectType>)
          - arg: NameNode(type=<CNumericType unsigned char>, name="elem")
```
The IteratorNode is `py_object_type`, should it be `bytearray_type`? The code suggests that `IteratorNode` implements `result = iter(sequence)`, when I do that in python I get a bytearray_iterator.
```python
>>> iter(bytearray(b'1234'))
<bytearray_iterator object at 0x7f1f941729a0>
```
  
Replacing the condition with `elif self.iterator.sequence.type is bytearray_type:` also seems to pass the tests, would that be better?

File: tests/run/bytearray_iteration_T3473.pyx [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+# ticket: 3473
+
+def test_bytearray_iteration(src):
+    """
+    >>> src = b'123'
+    >>> for elem in bytearray(src):
+    ...     print(elem)
+    49
+    50
+    51
```
[the remark this answers is not in the document]
Comment (Kenrick0) [reply tests/run/bytearray_iteration_T3473.pyx]: Cool, thanks

File: Cython/Compiler/ExprNodes.py (hunk 1 above)
Comment (da-woods) [reply Cython/Compiler/ExprNodes.py:2880]: > Replacing the condition with elif self.iterator.sequence.type is bytearray_type: also seems to pass the tests, would that be better?

Yes, I think so.

---------------------

> The master branch has a proper fix.

I kind of wonder if this is a general problem with any type that has a known, optimized index return type but no optimized for loop. Although at the moment the only example we know of is `bytearray`. I'm not sure whether we should worry about it or not - the fix in the master branch to optimize `bytearray` should make it very easy to optimize the loop for other types like this.

I'd probably go for the fix targetted specifically a `bytearray`, but I'm not completely convinced that this won't pop up again for some other type.

**Comment by Kenrick0** [comment]
It seems some of the travis checks are failing but they look unrelated to this change, is that expected?

**Comment by scoder** [comment]
Closing and reopening to get a clean CI run.

**Comment by scoder** [comment]
Hmm, travis-ci.org is gone, as advertised. Looks like we have to switch the `0.29.x` branch to GHA as well now. :-(

**Comment by scoder** [comment]
Triggering a clean CI build again.

**Comment by da-woods** [comment]
I think everyone's been happy with this for a while, it's just been overlooked. So I'm merging it.

Sorry for the delay and thanks for the PR.

**Pull request merged**